llvm/lib/CodeGen/TargetLoweringBase.cpp

```diff
@@ -2019,7 +2019,9 @@ void TargetLoweringBase::insertSSPDeclarations(Module &M) const {
     if (M.getDirectAccessExternalData() &&
         !TM.getTargetTriple().isWindowsGNUEnvironment() &&
         !TM.getTargetTriple().isOSFreeBSD() &&
-        !TM.getTargetTriple().isOSDarwin())
+        (!TM.getTargetTriple().isOSDarwin() ||
+         (TM.getTargetTriple().isARM() &&
```

Comment:
This ensures that code generation for arm*-apple-darwin -fno-pic does not change.

If `arm*-apple-darwin -fno-pic` is ok to use indirection, we can clean up this afterwards.

Reply:
Does the current version of the patch actually change the logic for MachO targets outside of this change?  I can't find any logical differences (the ARMInstrInfo.cpp changes are guarded by isTargetELF(), and the ARMBaseInstrInfo.cpp change is unreachable due to an earlier isTargetMachO()).

Reply:
I believe this patch does not change Mach-O behavior.

This `llvm/lib/CodeGen/TargetLoweringBase.cpp` change can also be removed and the behavior does not change. Sorry for including it here. However, the file is somewhat related.

I've added a commit to this PR and this now becomes
```
    if (M.getDirectAccessExternalData() &&
        !TM.getTargetTriple().isWindowsGNUEnvironment() &&
        !TM.getTargetTriple().isOSFreeBSD() &&
        (!TM.getTargetTriple().isOSDarwin() ||
         TM.getRelocationModel() == Reloc::Static))
      GV->setDSOLocal(true);
  }
```

This `TM.getRelocationModel() == Reloc::Static` condition refines 1ec30106a557b2c63d6d6c6b93634674aa65cdb9 .

I think Mach-O's static relocation model largely ignores `dso_local` IR markers and relies on `TM.shouldAssumeDSOLocal` to return true for static relocation code model. However, it seems more appropriate to me to encode this information correctly, which can probably enable code share between ELF and Mach-O.